src/particles/wakefields/ExecuteWakefield.H

```diff
@@ -0,0 +1,150 @@
+
+            int const num_bins = csr_bins;  // Set resolution
+            amrex::Real const bin_min = t_min;
+            amrex::Real const bin_max = t_max;
+            amrex::Real const bin_size = (bin_max - bin_min) / (num_bins - 1);  // number of evaluation points
+
+            // Allocate memory for the charge profile
+            amrex::Gpu::DeviceVector<amrex::Real> charge_distribution(num_bins + 1, 0.0);
+            amrex::Gpu::DeviceVector<amrex::Real> mean_x(num_bins, 0.0);
+            amrex::Gpu::DeviceVector<amrex::Real> mean_y(num_bins, 0.0);
+
+            // Call charge deposition function
+            impactx::particles::wakefields::DepositCharge1D(particle_container, charge_distribution, bin_min, bin_size, is_unity_particle_weight);
+
+            // Sum up all partial charge histograms to one MPI process to calculate the global wakefield.
+            // Once calculated, we will distribute convoluted_wakefield back to every MPI process.
+            amrex::ParallelReduce::Sum(
+                charge_distribution.data(),
+                charge_distribution.size(),
+                amrex::ParallelDescriptor::IOProcessorNumber(),
+                amrex::ParallelDescriptor::Communicator()
+            );
+
+            amrex::Gpu::DeviceVector<amrex::Real> convoluted_wakefield;
+            if (amrex::ParallelDescriptor::IOProcessor()) {
+                // Call the mean transverse position function
+                impactx::particles::wakefields::MeanTransversePosition(particle_container, mean_x, mean_y, bin_min,
+                                                                       bin_size, is_unity_particle_weight);
+
+                // Call charge density derivative function
+                amrex::Gpu::DeviceVector<amrex::Real> slopes(charge_distribution.size() - 1, 0.0);
+                impactx::particles::wakefields::DerivativeCharge1D(charge_distribution, slopes, bin_size,
+                                                                   GetNumberDensity); // Use number derivatives for convolution with CSR
+
+                // Construct CSR wake function on 2N support
+                amrex::Gpu::DeviceVector<amrex::Real> wake_function(num_bins*2, 0.0);
+                amrex::Real *const dptr_wake_function = wake_function.data();
+                auto const dR = R;  // for NVCC capture
+                amrex::ParallelFor(num_bins*2, [=] AMREX_GPU_DEVICE(int i) {
+                    if (i < num_bins) {
+                        amrex::Real const s = static_cast<amrex::Real>(i) * bin_size;
+                        dptr_wake_function[i] = impactx::particles::wakefields::w_l_csr(s, dR, bin_size);
+                    }
+                    else if (i > num_bins) {
+                        amrex::Real const s = static_cast<amrex::Real>(i - 2*num_bins) * bin_size;
+                        dptr_wake_function[i] = impactx::particles::wakefields::w_l_csr(s, dR, bin_size);
+                    }
+                });
```

Comment:
@xbojanic @cemitch99 I cannot yet spot the update for our integration constant at `i=0`. Is that missing?

Reply:
Oh, you treat that directly inside w_l_csr
https://github.com/ECP-WarpX/impactx/pull/627/commits/01422aa1fce947d35073608967577ad4d7640104